jakarta-eclipse/src/org/jakarta/jdt/PersistenceEntityDiagnosticsCollector.java

```diff
@@ -0,0 +1,181 @@
+		Integer isPrivateFinal = flag - Flags.AccPrivate;
+		Integer isProtectedFinal = flag - Flags.AccProtected;
+		Integer isFinal = flag;
+		if (isPublicFinal.equals(Flags.AccFinal) || 
+				isPrivateFinal.equals(Flags.AccFinal) || 
+				isProtectedFinal.equals(Flags.AccFinal) ||
+				isFinal.equals(Flags.AccFinal))
+		{
+			return true;
+		}
+		return false;
+	}
+	
+	private Diagnostic createDiagnostic(IJavaElement el, ICompilationUnit unit, String msg) {
+		try {
+			ISourceRange nameRange = JDTUtils.getNameRange(el);
+			Range range = JDTUtils.toRange(unit, nameRange.getOffset(), nameRange.getLength());
+			return new Diagnostic(range, msg);
+		} catch(JavaModelException e) {
+			Activator.logException("Cannot calculate diagnostics", e);
+		}
+		return null;
+	}
+
+	@Override
+	public void collectDiagnostics(ICompilationUnit unit, List<Diagnostic> diagnostics) {
+		// TODO Auto-generated method stub
+		if (unit != null) {
+			IType[] alltypes;
+			IAnnotation[] allAnnotations; 
+			try {
+				alltypes = unit.getAllTypes();
+				for (IType type: alltypes) {
+					String className = type.getElementName();
+					
+					allAnnotations = type.getAnnotations();
+					
+					ISourceRange nameRange = JDTUtils.getNameRange(type);
+					Range range = JDTUtils.toRange(unit, nameRange.getOffset(), nameRange.getLength());
+					
+					/*============ Entity Annotation Diagnostics ===========*/
+					IAnnotation EntityAnnotation = null;
+					for (IAnnotation annotation: allAnnotations) {
+						if (annotation.getElementName().equals(PersistenceConstants.ENTITY)) {
+							EntityAnnotation = annotation;
+						}
+					}
+					
+					if (EntityAnnotation != null) {
+						
+						ISourceRange annotationNameRange = JDTUtils.getNameRange(EntityAnnotation);
+						Range annotationrange = JDTUtils.toRange(unit, annotationNameRange.getOffset(), annotationNameRange.getLength());
+										
+						// Define boolean requirements for the diagnostics
+						boolean hasPublicOrProtectedNoArgConstructor = false;
+						boolean isEntityClassFinal = false;
+						boolean isMethodsOrPersistentVariablesFinal = false;
+						
+						// Get the Methods of the annotated Class
+						for (IMethod method: type.getMethods()) {
+							if (method.isConstructor()) {
+								// We have found a method that is a constructor
+								if (method.getNumberOfParameters() > 0) continue;
+								
+								// Don't need to perform subtractions to check flags because eclipse notifies on illegal constructor modifiers
+								if (method.getFlags() != Flags.AccPublic && method.getFlags() != Flags.AccProtected) continue;
+								
+								hasPublicOrProtectedNoArgConstructor = true;
+							}
+							
+							// All Methods of this class should not be final
+							if (this.isFinal(method.getFlags())) {
+								diagnostics.add(createDiagnostic(method, unit, "Class with @Entity annotation cannot contain any final methods or persistent instance variables."));
```

Comment:
This message should only mention methods:
"A class using the `@Entity` annotation cannot contain any methods that are declared final"